trustmanager/keydbstore.go

```diff
@@ -0,0 +1,153 @@
+package trustmanager
+
+import (
+	"database/sql"
+	"fmt"
+	"sync"
+
+	"github.com/docker/notary/pkg/passphrase"
+	"github.com/endophage/gotuf/data"
+	"github.com/jinzhu/gorm"
+	gojose "github.com/square/go-jose"
+)
+
+// KeyDBStore persists and manages private keys on a SQL database
+type KeyDBStore struct {
+	sync.Mutex
+	db         gorm.DB
+	passphrase string
+	encrypter  gojose.Encrypter
+	cachedKeys map[string]data.PrivateKey
+}
+
+// GormPrivateKey represents a PrivateKey in the database
+type GormPrivateKey struct {
+	gorm.Model
+	KeyID      string `sql:"not null;unique"`
+	Encryption string `sql:"not null"`
+	Algorithm  string `sql:"not null"`
+	Public     []byte `sql:"not null"`
+	Private    string `sql:"not null"`
+}
+
+// TableName sets a specific table name for our GormPrivateKey
+func (g GormPrivateKey) TableName() string {
+	return "private_keys"
+}
+
+// NewKeyDBStore returns a new KeyDBStore backed by a SQL database
+func NewKeyDBStore(passphraseRetriever passphrase.Retriever, dbType string, dbSQL *sql.DB) (*KeyDBStore, error) {
+	cachedKeys := make(map[string]data.PrivateKey)
+
+	// Retreive the passphrase that will be used to encrypt the keys
+	passphrase, _, err := passphraseRetriever("", "", false, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	// Setup our encrypted object
+	encrypter, err := gojose.NewEncrypter(gojose.A256GCMKW, gojose.A256GCM, []byte(passphrase))
```

Comment:
I'm not sure we should be using A128GCMKW here. This code will use the passphrase directly as the encryption key and do no key derivation. Hilariously (or maybe intentionally?), the random garbage you typed in the test is exactly 32 bytes which means it works as a 256 bit AES key.

If we're going to use passphrase for key encryption we need to be using something like PBES2-HS256+A128KW from https://tools.ietf.org/html/draft-miller-jose-jwe-protected-jwk-02#section-6.